internal/functions/efgs/upload-batch.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/covid19cz/erouska-backend/internal/secrets"
+	"github.com/golang/protobuf/proto"
+	"io/ioutil"
+	"net/http"
+	urlutils "net/url"
+	"os"
+	"strconv"
+	"time"
+)
+
+//UploadBatch Called in CRON every 2 hours. Gets keys from database and upload them to EFGS.
+func UploadBatch(w http.ResponseWriter, r *http.Request) {
+	var ctx = r.Context()
+	logger := logging.FromContext(ctx)
+	currentTime := time.Now()
+	timeFrom := currentTime.AddDate(0, 0, -14).Format("2006-01-02")
+	maxBatchSize, isSet := os.LookupEnv("EFGS_UPLOAD_BATCH_SIZE")
+	if !isSet {
+		logger.Error("Maximum batch size not set")
+		return
+	}
+
+	batchSizeLimit, err := strconv.Atoi(maxBatchSize)
+	if err != nil {
+		logger.Errorf("Error converting batch size to int: %s", err)
+	}
+
+	keys, err := database.GetDiagnosisKeys(timeFrom)
+	if err != nil {
+		logger.Errorf("Downloading keys error: %s", err)
+	}
+
+	if len(keys) <= 0 {
+		logger.Warn("No new keys in database.")
+		return
+	}
+
+	var diagnosisKeys []*DiagnosisKey
+	for _, k := range keys {
+		diagnosisKeys = append(diagnosisKeys, k.ToData())
+	}
+
+	sortDiagnosisKey(diagnosisKeys)
+	uploadConfig, err := uploadBatchConfiguration(ctx)
+	if err != nil {
+		logger.Errorf("Upload configuration error: %v", err)
+		return
+	}
+
+	batches := splitBatch(diagnosisKeys, batchSizeLimit)
+	for _, batch := range batches {
+		diagnosisKeyBatch := makeBatch(batch)
+		hash := sha1.New()
+		hash.Write(batchToBytes(&diagnosisKeyBatch))
+		uploadConfig.BatchTag = currentTime.Format("20060102") + "-" + hex.EncodeToString(hash.Sum(nil))[:7]
+		logger.Debugf("Uploading batch (%d keys) with tag %s", len(batch), uploadConfig.BatchTag)
+
+		resp, err := uploadBatch(ctx, &diagnosisKeyBatch, uploadConfig)
+		if err != nil {
+			logger.Errorf("Batch upload failed: %v", err)
+			return
+		}
+		if resp != nil {
+			filterInvalidDiagnosisKeys(ctx, resp, keys)
+		}
+		logger.Debugf("Batch %s successfully uploaded", uploadConfig.BatchTag)
+	}
+
+	if err := database.RemoveDiagnosisKey(keys); err != nil {
+		logger.Errorf("Removing uploaded keys from database failed: %s", err)
+		return
+	}
+
+	logger.Infof("%d keys (in %d batches) successfully uploaded", len(keys), len(batches))
+}
+
+func uploadBatchConfiguration(ctx context.Context) (*uploadConfiguration, error) {
+	logger := logging.FromContext(ctx)
+	secretsClient := secrets.Client{}
+
+	var err error
+	var config uploadConfiguration
+
+	config.NBTLSPair, err = efgsutils.LoadX509KeyPair(ctx, efgsutils.NBTLS)
+	if err != nil {
+		logger.Debug("Error loading authentication certificate")
+		return nil, err
+	}
+
+	config.Client, err = efgsutils.NewEFGSClient(ctx, config.NBTLSPair)
+	if err != nil {
+		logger.Debug("Could not create EFGS client")
+		return nil, err
+	}
+
+	// TODO remove while switching to real efgs
+	_, connectToLocal := os.LookupEnv("EFGS_LOCAL")
+
+	var rootURL []byte
+	if connectToLocal {
+		rootURL, err = secretsClient.Get("efgs-root-url")
```

Comment:
Ehm, I'm ashamed I copied it even with my copy-paste error :smile: Here it should be `efgs-test-url` instead.